js/cards/smartport.ts

```diff
@@ -0,0 +1,460 @@
+/* Copyright 2010-2019 Will Scullin <scullin@scullinsteel.com>
+ *
+ * Permission to use, copy, modify, distribute, and sell this software and its
+ * documentation for any purpose is hereby granted without fee, provided that
+ * the above copyright notice appear in all copies and that both that
+ * copyright notice and this permission notice appear in supporting
+ * documentation.  No representations are made about the suitability of this
+ * software for any purpose.  It is provided "as is" without express or
+ * implied warranty.
+ */
+
+import { debug, toHex } from '../util';
+import { rom } from '../roms/cards/smartport';
+import { Card, Restorable, byte, word } from '../types';
+import CPU6502, { CpuState } from '../cpu6502';
+
+type SmartDisk = Uint8Array[];
+
+interface BlockDevice {
+    blocks: Uint8Array
+}
+
+export interface SmartPortState {
+    disks: SmartDisk[]
+}
+
+export interface SmartPortOptions {
+    block: boolean;
+}
+
+class Address {
+    lo: byte;
+    hi: byte;
+
+    constructor(private cpu: CPU6502, a: byte | word, b?: byte) {
+        if (b === undefined) {
+            this.lo = a & 0xff;
+            this.hi = a >> 8;
+        } else {
+            this.lo = a;
+            this.hi = b;
+        }
+    }
+
+    loByte() {
+        return this.lo;
+    }
+
+    hiByte() {
+        return this.hi;
+    }
+
+    inc(val: byte) {
+        return new Address(this.cpu, ((this.hi << 8 | this.lo) + val) & 0xffff);
+    }
+
+    readByte() {
+        return this.cpu.read(this.hi, this.lo);
+    }
+
+    readWord() {
+        const readLo = this.readByte();
+        const readHi = this.inc(1).readByte();
+
+        return readHi << 8 | readLo;
+    }
+
+    readAddress() {
+        const readLo = this.readByte();
+        const readHi = this.inc(1).readByte();
+
+        return new Address(this.cpu, readLo, readHi);
+    }
+
+    writeByte(val: byte) {
+        this.cpu.write(this.hi, this.lo, val);
+    }
+
+    writeWord(val: word) {
+        this.writeByte(val & 0xff);
+        this.inc(1).writeByte(val >> 8);
+    }
+
+    writeAddress(val: Address) {
+        this.writeByte(val.loByte());
+        this.inc(1).writeByte(val.hiByte());
+    }
+
+    toString() {
+        return '$' + toHex(this.hi) + toHex(this.lo);
+    }
+}
+
+export default class SmartPort implements Card, Restorable<SmartPortState> {
+    COMMAND = 0x42;
+    UNIT = 0x43;
+    ADDRESS_LO = 0x44;
+    // const ADDRESS_HI = 0x45;
+    BLOCK_LO = 0x46;
+    // const BLOCK_HI = 0x47;
+
+    private rom: Uint8Array;
+    private disks: SmartDisk[] = [];
+
+    constructor(private cpu: CPU6502, options: SmartPortOptions) {
+        this.rom = new Uint8Array(rom);
```

Comment:
Where is `rom` coming from here? Do you want to use `ROM` from 'rom.ts'?

Reply:
It's coming from `../roms/cards/smartport`. I should probably go in and de-dupe that name.